lib/fluent/plugin/out_secondary_file.rb

```diff
@@ -0,0 +1,133 @@
+
+    FILE_PERMISSION = 0644
+    DIR_PERMISSION = 0755
+    PLACEHOLDER_REGEX = /\${(tag(\[\d+\])?|[\w.@-]+)}/
+
+    desc "The directory path of the output file."
+    config_param :directory, :string, default: "secondary/"
+    desc "The baseanme of the output file."
+    config_param :basename, :string, default: "dump.bin"
+    desc "The flushed chunk is appended to existence file or not."
+    config_param :append, :bool, default: false
+    config_param :compress, :enum, list: [:text, :gz, :gzip], default: :text
+
+    def configure(conf)
+      super
+
+      unless @as_secondary
+        raise Fluent::ConfigError, "This plugin can only be used in the <secondary> section"
+      end
+
+      @placeholders = @basename.scan(PLACEHOLDER_REGEX).flat_map(&:first) # to trim suffix [\d+]
+      validate_basename_is_compatible_with_primary_buffer!
+
+      @directory += "/" if @directory.size != 0 && !@directory.end_with?("/")
+
+      test_path = generate_path(Time.now.strftime("%Y%m%d"))
+      unless Fluent::FileUtil.writable_p?(test_path)
+        raise Fluent::ConfigError, "out_secondary_file: `#{test_path}` is not writable"
+      end
+
+      @dir_perm = system_config.dir_permission || DIR_PERMISSION
+      @file_perm = system_config.file_permission || FILE_PERMISSION
+    end
+
+    def write(chunk)
+      name = extract_placeholders(@basename, chunk.metadata)
+      path = generate_path(name)
+      FileUtils.mkdir_p File.dirname(path), mode: @dir_perm
+
+      case @compress
+      when :text
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          chunk.write_to(f)
+        }
+      when :gz, :gzip
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          gz = Zlib::GzipWriter.new(f)
+          chunk.write_to(gz)
+          gz.close
+        }
+      end
+
+      path
+    end
+
+    # To avoid calling Fluent::Compat::TimeSlicedOutput#extract_placeholders if primary plugin is `out_file`
+    def extract_placeholders(str, metadata)
+      super
+    end
+
+    private
+
+    def validate_basename_is_compatible_with_primary_buffer!
+      if !@chunk_key_time && path_has_time_format?
+        raise Fluent::ConfigError, "out_secondary_file: File path has an incompatible placeholder, add time formats, like `%Y%m%d`, to `path`"
+      end
+
+      if !@chunk_key_tag && (ph = @placeholders.find { |p| p.match(/tag(\[\d+\])?/) })
+        raise Fluent::ConfigError, "out_secondary_file: File path has an incompatible placeholder #{ph}, add tag placeholder, like `${tag}`, to `path`"
+      end
+
+      if @chunk_keys.empty?
+        vars = @placeholders.reject { |p| p.match(/tag(\[\d+\])?/) }
+
+        if ph = vars.find { |v| !@chunk_keys.include?(v) }
+          raise Fluent::ConfigError, "out_secondary_file: File path has an incompatible placeholder #{ph}, add variable placeholder, like `${varname}`, to `path`"
+        end
+      end
```

Comment:
What is this `if` block checking?
If `@chunk_keys.empty?` is true, `!@chunk_keys.include?(v)` is always true.
Check both of code and test code.